arlpy/bf.py

```diff
@@ -271,7 +272,69 @@ def capon(x, fc, sd, complex_output=False):
 
-def broadband(x, fs, nfft, sd, f0=0, fmin=None, fmax=None, overlap=0, beamformer=bartlett):
+def music(x, fc, sd, complex_output=False,num_sig_components = 1):
+    """Frequency-domain MUSIC beamformer.
+
+    The timeseries data must be 2D with narrowband complex timeseries for each sensor in
+    individual rows. The steering delays must also be 2D with a row per steering direction.
+
+    If the timeseries data is specified as 1D array, it is assumed to represent multiple sensors
+    at a single time.
+
+    The covariance matrix of x is estimated over the entire timeseries, and used to compute
+    the optimal weights for the MUSIC beamformer via eigen-decomposition.
+
+    :param x: narrowband complex timeseries data for multiple sensors (row per sensor)
+    :param fc: carrier frequency for the array data (Hz)
+    :param sd: steering delays (s)
+    :param complex_output: True for complex signal, False for beamformed power
+    :param num_sig_components: Number of signal eigenvectors (rest are considered noise)
+    :returns: beamformer output averaged across time
+
+    >>> from arlpy import bf
+    >>> import numpy as np
+    >>> # narrowband (1 kHz) timeseries array data assumed to be loaded in x
+    >>> # sensor positions assumed to be in pos
+    >>> y = bf.music(x, 1000, bf.steering(pos, 1500, np.linspace(-np.pi/2, np.pi/2, 181)))
+    """
+    if x.ndim == 1:
+        x = x[:,_np.newaxis]
+    if x.shape[0] != sd.shape[1]:
+        raise ValueError('Sensor count mismatch in data and steering vector')
+    if fc == 0:
+        a = _np.ones_like(sd)
+    else:
+        a = _np.exp(-2j*_np.pi*fc*sd)/_np.sqrt(sd.shape[1])
+    if complex_output:
+        R = covariance(x)
+        if _np.linalg.cond(R) > 10000:
+            R += _np.random.normal(0, _np.max(_np.abs(R))/1000000, R.shape)
+        A, B = _LA.eigh(R)
+        idx = A.argsort()[::-1]
+        lmbd = A[idx] # Sorted vector of eigenvalues
+        B = B[:, idx] # Eigenvectors rearranged accordingly
+        En = B[:, num_sig_components:len(B)] # Noise eigenvectors
+
+        V = _np.matmul(En,En.conj().T)#En.conj().dot(En)#
+        w = _np.array([V.dot(a[j])/(a[j].conj().dot(V).dot(a[j])) for j in range(a.shape[0])])
+        return w.conj().dot(x)
+
+    else:
+        R = covariance(x)
+        if _np.linalg.cond(R) > 10000:
+            R += _np.random.normal(0, _np.max(_np.abs(R))/1000000, R.shape)
+
+        A, B = _LA.eigh(R)
+        idx = A.argsort()[::-1]
+        lmbd = A[idx] # Sorted vector of eigenvalues
+        B = B[:, idx] # Eigenvectors rearranged accordingly
+        En = B[:, num_sig_components:len(E)] # Noise eigenvectors
+
+        V = _np.matmul(En,En.conj().T)
+
+        return _np.array([1.0/a[j].conj().dot(V).dot(a[j]).real for j in range(a.shape[0])])
+
+def broadband(x, fs, nfft, sd, f0=0, fmin=None, fmax=None, overlap=0, beamformer=bartlett,num_sig_components = 1):
```

Comment:
I am not so comfortable with adding MUSIC specific parameters in the `broadband()` function. How about adding arbitrary keyword parameters at the end, which are verbatim passed to the underlying beamformer? That way we can pass this to MUSIC, but also any other parameters to any other beamformer. Open to this change?

Reply:
Agreed. This was a quick-fix I added, mainly to highlight that this needs to be done. I too was thinking of what might be the best way to do this broadly in a way that can be used for all beamformers.

Reply:
If you feel comfortable making this change, do it. Otherwise PR without the change, and I can make the change when merging. Either way works for me.